crates/ruff_python_formatter/src/expression/binary_like.rs

```diff
@@ -20,13 +20,182 @@ use crate::expression::string::StringLayout;
+            }));
+
+            rec(
+                Operand::Right {
+                    expression: binary.right.as_ref(),
+                    trailing_comments,
+                },
+                comments,
+                source,
+                parts,
+            )
+        }
+
+        fn rec<'a>(
+            operand: Operand<'a>,
+            comments: &'a Comments,
+            source: &str,
+            parts: &mut SmallVec<[OperandOrOperator<'a>; 8]>,
+        ) -> FormatResult<()> {
+            let expression = operand.expression();
+            match expression {
+                Expr::BinOp(binary) if !is_expression_parenthesized(expression.into(), source) => {
+                    let leading_comments = operand
+                        .leading_binary_comments()
+                        .unwrap_or_else(|| comments.leading(binary));
+
+                    let trailing_comments = operand
+                        .trailing_binary_comments()
+                        .unwrap_or_else(|| comments.trailing(binary));
+
+                    recurse_binary(
+                        binary,
+                        leading_comments,
+                        trailing_comments,
+                        comments,
+                        source,
+                        parts,
+                    )
+                }
+                Expr::Compare(compare)
+                    if !is_expression_parenthesized(expression.into(), source) =>
+                {
+                    let leading_comments = operand
+                        .leading_binary_comments()
+                        .unwrap_or_else(|| comments.leading(compare));
+
+                    let trailing_comments = operand
+                        .trailing_binary_comments()
+                        .unwrap_or_else(|| comments.trailing(compare));
+
+                    recurse_compare(
+                        compare,
+                        leading_comments,
+                        trailing_comments,
+                        comments,
+                        source,
+                        parts,
+                    )
+                }
+                _ => {
+                    parts.push(OperandOrOperator::Operand(operand));
+                    Ok(())
+                }
+            }
+        }
+
+        let mut parts = SmallVec::new();
+        match self {
+            BinaryLike::BinaryExpression(binary) => {
+                // Leading and trailing comments are handled by the binary's ``FormatNodeRule` implementation.
+                recurse_binary(binary, &[], &[], comments, source, &mut parts)?;
+            }
+            BinaryLike::CompareExpression(compare) => {
+                // Leading and trailing comments are handled by the compare's ``FormatNodeRule` implementation.
+                recurse_compare(compare, &[], &[], comments, source, &mut parts)?;
+            }
+        }
+
+        Ok(FlatBinaryExpression(parts))
```

Comment:
Could you add some comments about what datastructure we're building here?

Reply:
There's somewhat extensive documentation on the `FlatBinaryExpressionSlice`

Reply:
Could you link `FlatBinaryExpressionSlice` here?